src/extra/recipes/cookiebox.py

```diff
@@ -0,0 +1,1047 @@
+        """
+        # base properties
+        self._run = run
+        self._energy_axis = energy_axis
+        self._tof_settings = tof_settings
+
+        self._auger_start_roi = {tof_id: self._init_auger_start_roi for tof_id in self._tof_settings.keys()}
+        self._start_roi = {tof_id: self._init_start_roi for tof_id in self._tof_settings.keys()}
+        self._stop_roi = {tof_id: self._init_stop_roi for tof_id in self._tof_settings.keys()}
+        self._filter_length = self._init_filter_length
+        if not isinstance(self._filter_length, dict):
+            self._filter_length = {tof_id: self._filter_length for tof_id in self._tof_settings.keys()}
+
+        # now do the full analysis, step by step
+        # get XGM and auxiliary data
+        self.update_xgm_and_metadata()
+
+        # update eTOF data reading objects
+        self.update_tof_settings()
+
+        # find RoI if needed
+        self.update_roi()
+
+        # find where the peaks are per energy in each Tof
+        self.update_fit_result()
+
+        # calibrate and calculate transmission
+        self.update_calibration()
+
+        # now we can use the apply method
+        logging.info("Ready to apply energy calibration and transmission correction to analysis data ...")
+
+    # these setters allow one to update
+    # some of the input properties and rerun only the necessary steps
+    # for recalibration
+    #
+    @property
+    def run(self) -> DataCollection:
+        return self._run
+
+    @run.setter
+    def run(self, value: DataCollection):
+        self._run = value
+        self.update_xgm_and_metadata()
+        self.update_tof_settings()
+        self.update_roi()
+        self.update_fit_result()
+        self.update_calibration()
+
+    @property
+    def first_pulse_offset(self) -> int:
+        return self._first_pulse_offset
+
+    @first_pulse_offset.setter
+    def first_pulse_offset(self, value: int):
+        self._first_pulse_offset = value
+        self.update_tof_settings()
+        self.update_roi()
+        self.update_fit_result()
+        self.update_calibration()
+
+    @property
+    def single_pulse_length(self) -> int:
+        return self._single_pulse_length
+
+    @single_pulse_length.setter
+    def single_pulse_length(self, value: int):
+        self._single_pulse_length = value
+        self.update_tof_settings()
+        self.update_roi()
+        self.update_fit_result()
+        self.update_calibration()
+
+    @property
+    def interleaved(self) -> bool:
+        return self._interleaved
+
+    @interleaved.setter
+    def interleaved(self, value: bool):
+        self._interleaved = value
+        self.update_tof_settings()
+        self.update_roi()
+        self.update_fit_result()
+        self.update_calibration()
+
+    @property
+    def xgm_source(self) -> str:
+        return self._xgm_source
+
+    @xgm_source.setter
+    def xgm_source(self, value: str):
```

Comment:
There's a lot of these getters and setters :sweat_smile: I'm not sure how this class is typically used, but would it make sense to require that some/all parameters are passed in the constructor, and changing any of them requires creating a new object? In that sense the class would become a bit more 'read-only', which I think is simpler than allowing mutation everywhere.

Reply:
Your suggestion was my original version and I agree it is simpler ... It has been proposed that I could find a scheme to avoid redoing it all. In practice, it is indeed annoying to re-build the full object if you are optimizing parameters (when the autodetection does not work).